File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    int[] numbers;
    int target;

    public int solution(int[] numbers, int target) {
        this.numbers = numbers;
        this.target = target;

        return dfs(0, 0, 0);
    }

    private int dfs(int index, int value, int cnt) {
        // index가 끝까지 도달했는지 먼저 체크하기 때문에 
        // 배열의 원소를 모두 연산하기 위해 index가 lastIndex + 1 (= length) 일때 값이 target과 값이 같은지 체크  
        if (index == this.numbers.length) {
            if (value == this.target) {
                cnt++;
            }
        } else {
            // 더하기
            cnt = dfs(index + 1, value + this.numbers[index], cnt);

            // 빼기
            cnt = dfs(index + 1, value - this.numbers[index], cnt);
        }

        return cnt;
    }
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.kt
class Solution {
    fun solution(numbers: IntArray, target: Int): Int {
        return dfs(numbers, target, 0, 0, 0)
    }

    private fun dfs(numbers: IntArray, target: Int, index: Int, value: Int, cnt: Int): Int {
        var _cnt: Int = cnt

        if (index == numbers.size) {
            if (value == target) {
                _cnt++
            }
        } else {
            // 더하기
            _cnt = dfs(numbers, target, index + 1, value + numbers[index], _cnt)

            //빼기
            _cnt = dfs(numbers, target, index + 1, value - numbers[index], _cnt)
        }

        return _cnt
    }
}